mqtt/mqtt-bridge/src/client.rs

```diff
@@ -0,0 +1,223 @@
+#![allow(dead_code)] // TODO remove when ready
+use std::{collections::HashSet, time::Duration};
+
+use anyhow::Result;
+use async_trait::async_trait;
+use chrono::Utc;
+use futures_util::future::BoxFuture;
+use tokio::{net::TcpStream, stream::StreamExt};
+use tracing::{debug, error, warn};
+
+use mqtt3::{
+    proto, Client, Event, IoSource, ShutdownError, SubscriptionUpdateEvent, UpdateSubscriptionError,
+};
+
+use crate::settings::Credentials;
+use crate::token_source::{SasTokenSource, TokenSource};
+
+const DEFAULT_TOKEN_DURATION_SECS: Duration = Duration::from_secs(60 * 60);
+
+#[derive(Debug)]
+pub struct ShutdownHandle(mqtt3::ShutdownHandle);
+
+impl ShutdownHandle {
+    pub async fn shutdown(&mut self) -> Result<(), ClientConnectError> {
+        self.0
+            .shutdown()
+            .await
+            .map_err(ClientConnectError::ShutdownClient)?;
+        Ok(())
+    }
+}
+
+pub struct TcpConnection<T>
+where
+    T: TokenSource + Clone + Send + Sync + 'static,
+{
+    address: String,
+    token_source: Option<T>,
+}
+
+impl<T> IoSource for TcpConnection<T>
+where
+    T: TokenSource + Clone + Send + Sync + 'static,
+{
+    type Io = TcpStream;
+    type Error = std::io::Error;
+    type Future = BoxFuture<'static, Result<(TcpStream, Option<String>), std::io::Error>>;
+
+    fn connect(&mut self) -> Self::Future {
+        let address = self.address.clone();
+        let token_source = self.token_source.as_ref().cloned();
+
+        Box::pin(async move {
+            let expiry = Utc::now()
+                + chrono::Duration::from_std(DEFAULT_TOKEN_DURATION_SECS)
+                    .unwrap_or_else(|_| chrono::Duration::hours(1));
+            let mut password = None;
+            if let Some(ts) = token_source {
+                password = match ts.get(&expiry).await {
+                    Ok(x) => Some(x),
+                    Err(e) => {
+                        warn!("Failed to get token {}", e);
+                        None
+                    }
+                }
+            };
+
+            let io = TcpStream::connect(address).await;
+            io.map(|io| (io, password))
+        })
+    }
+}
+
+pub struct MqttClient<T>
```

Comment:
nit: it might be confusing to have Client and MqttClient types. Also, it is just a wrapper. We could maybe indicate that in the name? I don't think it's too bad if we leave it as-is though.